src/internal/connector/onedrive/collections.go

```diff
@@ -469,23 +464,132 @@ func updateCollectionPaths(
 			continue
 		}
 
-		ocol, ok := c.(*Collection)
-		if !ok {
-			return found, clues.New("unable to cast onedrive collection")
-		}
-
 		colPath := c.FullPath()
 
 		// Only updates if initialCurPath parent of colPath
 		updated := colPath.UpdateParent(initialCurPath, curPath)
 		if updated {
-			ocol.SetFullPath(colPath)
+			c.SetFullPath(colPath)
 		}
 	}
 
 	return found, nil
 }
 
+func (c *Collections) handleDelete(
+	itemID, driveID string,
+	oldPaths, newPaths map[string]string,
+	isFolder bool,
+	excluded map[string]struct{},
+) error {
+	if !isFolder {
+		excluded[itemID+DataFileSuffix] = struct{}{}
+		excluded[itemID+MetaFileSuffix] = struct{}{}
+		// Exchange counts items streamed through it which includes deletions so
+		// add that here too.
+		c.NumFiles++
+		c.NumItems++
+
+		return nil
+	}
+
+	var prevPath path.Path
+
+	prevPathStr, ok := oldPaths[itemID]
+	if ok {
+		var err error
+
+		prevPath, err = path.FromDataLayerPath(prevPathStr, false)
+		if err != nil {
+			return clues.Wrap(err, "invalid previous path").
+				With("collection_id", itemID, "path_string", prevPathStr)
+		}
+	}
+
+	// Nested folders also return deleted delta results so we don't have to
+	// worry about doing a prefix search in the map to remove the subtree of
+	// the deleted folder/package.
+	delete(newPaths, itemID)
+
+	if prevPath == nil {
+		// It is possible that an item was created and
+		// deleted between two delta invocations. In
+		// that case, it will only produce a single
+		// delete entry in the delta response.
+		return nil
+	}
+
+	col := NewCollection(
+		c.itemClient,
+		nil,
+		prevPath,
+		driveID,
+		c.service,
+		c.statusUpdater,
+		c.source,
+		c.ctrl,
+		// DoNotMerge is not checked for deleted items.
+		false,
+	)
+
+	c.CollectionMap[itemID] = col
+
+	return nil
+}
+
+func (c *Collections) getCollectionPath(
+	driveID string,
+	item models.DriveItemable,
+) (path.Path, error) {
+	var (
+		collectionPathStr string
+		itemID            = ptr.Val(item.GetId())
+		isNonRootFolder   = item.GetRoot() == nil &&
+			(item.GetFolder() != nil || item.GetPackage() != nil)
+	)
+
+	if item.GetRoot() != nil {
+		collectionPathStr = fmt.Sprintf(rootDrivePattern, driveID)
+	} else {
+		if item.GetParentReference() == nil ||
+			item.GetParentReference().GetPath() == nil {
+			err := clues.New("no parent reference").With("item_id", itemID)
+			if item.GetName() != nil {
+				err = err.With("item_name", ptr.Val(item.GetName()))
+			}
+
+			return nil, err
+		}
+
+		collectionPathStr = ptr.Val(item.GetParentReference().GetPath())
```

Comment:
Just wondering if we should just append the name here itself instead of after creating the path.

Reply:
yea, I debated about that for a bit too. I can add that if you feel strongly about it. The reason I ended up using `path.Append` below instead is we know at that point how to add items to a path because we control the path package. If we append the folder element here, we're making assumptions about what the format of the path is from Graph (mostly path separators). I know they won't change this cause it will break a lot of people's code, but part of me kind of likes to reduce assumptions made about stuff we don't have direct control over 😅